tools/nodejs/lib/duckdb.js

```diff
@@ -23,42 +23,130 @@ Connection.prototype.each = function(sql) {
     return statement.each.apply(statement, arguments);
 }
 
-Database.prototype.prepare = function() {
-    if (this.default_connection == undefined) {
-        this.default_connection = new duckdb.Connection(this);
+function ptr_to_arr(buffer, ptype, n) {
+    // TODO can we create those on the C++ side of things already?
+    switch(ptype) {
+        case 'BOOL':
+        case 'UINT8':
+            return new Uint8Array(buffer.buffer, 0, n);
+        case 'INT8':
+            return new Int8Array(buffer.buffer, 0, n);
+        case 'INT16':
+            return new Int16Array(buffer.buffer, 0, n);
+        case 'UINT16':
+            return new UInt16Array(buffer.buffer, 0, n);
+        case 'INT32':
+            return new Int32Array(buffer.buffer, 0, n);
+        case 'UINT32':
+            return new UInt32Array(buffer.buffer, 0, n);
+        case 'FLOAT':
+            return new Float32Array(buffer.buffer, 0, n);
+        case 'DOUBLE':
+            return new Float64Array(buffer.buffer, 0, n);
+        case 'VARCHAR':  // we already have created a string array on the C++ side for this
+            return buffer;
+        default:
+            return new Array<string>(0); // cough
     }
-    return this.default_connection.prepare.apply(this.default_connection, arguments);
 }
 
-Database.prototype.run = function() {
-    if (this.default_connection == undefined) {
-        this.default_connection = new Connection(this);
+// this follows the wasm udfs somewhat but is simpler because we can pass data much more cleanly
+Connection.prototype.register = function(name, return_type, fun) {
+    // TODO what if this throws an error somewhere? do we need a try/catch?
+    return this.register_bulk(name, return_type, function(descr, data_buffers) {
+        try {
+            const data_arr = [];
+            const validity_arr = [];
+
+            for (const idx in descr.args) {
+                const arg = descr.args[idx];
+                validity_arr.push(data_buffers[arg.validity_buffer]);
+                data_arr.push(ptr_to_arr(data_buffers[arg.data_buffer], arg.physical_type, descr.rows));
+            }
+
+            const out_data = ptr_to_arr(data_buffers[descr.ret.data_buffer], descr.ret.physical_type, descr.rows);
+            const out_validity = data_buffers[descr.ret.validity_buffer];
+
+            switch (descr.args.length) {
+                case 0:
+                    for (let i = 0; i < descr.rows; ++i) {
+                        const res = fun();
+                        out_data[i] = res;
+                        out_validity[i] = res == undefined ? 0 : 1;
```

Comment:
I'd prefer null over undefined here as return type for nulls.
Maybe `res === undefined || res === null` ?

Reply:
So what happens if we get undefined then?

Reply:
maybe `out_validity[i] = (res == undefined || res === null)? 0 : 1;` ?
I just thought that supporting an actual null as return value might be semantically better.